### ISSUE TITLE
Basic support for Named strings

### ISSUE THREAD

**Issue by zxcat** [Kozea/WeasyPrint#174]
It's quite simple to add. And very useful, e.g.:
```
h1 { string-set: title content() }
@page { @top-right { content: string(title) }}
```
Details: http://www.w3.org/TR/css3-gcpm/#running-headers-and-footers

Basic `content()` and `string(name)` support without any other parameters and forms will be great. _(Looks like other variants like `env()` are not mature enough to implement.)_

I think implementation is quite easy:
1. `string-set:` is much like `counter-reset:name value`. The only difference: `value` there is more complex, like `content:`;
2. `string(name)` is exactly like `counter(name)`, but with different dictionary.

I've tried to implement this myself, but I don't understand Python source.

Any chances *Named strings* will be in the next version of WeasyPrint? : )

**Comment by SirNeuman** [comment]
Is there any way around this for the time being? I, for example, need a string from my html page (a unique ID number) to be displayed in @bottom-left{}. So I wanted to do in the css:
 #ID-Number {string-set: the-number content();} 
@page{@bottom-left{content: string(the-number)}}
but the area at the bottom left of the pdf just shows up blank. I didn't realize I couldn't do this until I found this post, but is there some way of getting that ID-Number in the bottom left of the pdf some other way? If i try and put it in the html it appears above where i want it to appear, which is not in line with the rest of my footer (I have other stuff in bottom-center and bottom-right)

**Comment by SimonSapin** [comment]
@SirNeuman Your `#ID-Number` selector seems to imply that you have only one such element. If you want some content from the document in page footers *that is the same on all pages*, you can use `position: fixed` (which repeats on all pages) and some margin or positioning trick (like `top: 100%` maybe) to move it into the footer.

However if that number varies within a document (e.g. a chapter number), then the only way is named strings or some other feature very much like it.

**Comment by SirNeuman** [comment]
ahh ok. Thanks. Knowing that fixed position puts it on every page helps.

**Comment by E-M-P-I-R-E** [comment]
I am interested in working on this. If you are able to provide some guidance, @SimonSapin, that would be fantastic.

**Comment by SimonSapin** [comment]
@E-M-P-I-R-E Good!

Unfortunately, the GCPM spec suffers from being largely under-defined and having received very little review. For example, it’s not clear to me whether the `string()` function of the `content` property is valid in contexts other than on page-margin boxes (e.g. `::before` or `::after` pseudo-elements) and what the second argument means in this case.

If you feel extremely motivated, you could subscribe to the W3C CSS Working Group’s mailing list https://lists.w3.org/Archives/Public/www-style/ and provide feedback… But be prepared to spend *a lot* of your time for things to move very slowly.

But regardless, you should look at the latest version at http://dev.w3.org/csswg/css-gcpm/#named-strings rather than the dated snapshot on /TR.

You’ll want to start by adding the `string-set` property. In `weasyprint/css/properties.py`, add an entry for it in the `INITIAL_VALUES` dict. In `weasyprint/css/validation.py`, add a parsing function for its value with a `@validator(prefixed=True)` decorator. Note that (a subset of) `<content-list>` is already parsed for the `bookmark-label` property, so try to re-use that. In particular, the property currently only accepts one value, whereas the spec allows multiple space-separated values. If you want to make that subset more complete, do so in a way that is works for both property. But feel free to just keep the existing subset. You’ll also need to add parsing code for the `string()` function in the `content` property.

Next, you’ll want to resolve that property value to a single string of text for a given element. This is already done in the `resolve_bookmark_labels` function in `weasyprint/formatting_structure/build.py`, but the same function walks recursively over the tree and sets a `.bookmark_label` attribute on box objects, so extract the text resolution into a separate function that can be used for `string-set`.

Now, unlike `bookmark-label` or the `content` property on (pseudo-)elements, the difficulty here is that you want, for a given named string, a value potentially different on every page. Bookmarks and pseudo-elements are resolved during box construction (`weasyprint/formatting_structure/build.py`), but that doesn’t work here because box generation is not aware of pages.

(The box tree is what layout works on. Each element generates zero or more boxes, but usually just one. During layout, boxes are then split up into fragments that go onto different lines or pages, but WeasyPrint uses the same data structure for fragments as for boxes.)

So you’ll need to resolve named strings during layout. Fortunately, there is already a `LayoutContext` object with a single instance created for each document and passed through most of the layout code. I imagine you could have a dict mapping string names to their current value, and a function or method that updates this dict for a given box. The downside of doing this kind of thing in layout rather than box generation is that each type of box (float, table, ...) has its own layout code, so you’ll have to make sure that every one of them updates named strings appropriately.

Finally, add support for the `string()` function in `content_to_boxes()` in `weasyprint/formatting_structure/build.py`. `content_to_boxes()` should probably be modified to take an optional `context` argument (of type `LayoutContext`) that `make_margin_boxes()` in `weasyprint/layout/pages.py` would provide but other callers would not. This means that `string()` would not be supported on (pseudo-)elements, but arguably that’s not very useful anyway. Supporting that would require either moving resolution of the `content` property into layout rather than box generation, even for (pseudo-)elements; or doing resolution of named strings both during box generation and layout.

----

So, yeah. This is not a small feature to get started with. I hope you manage to make progress, and feel free to ask any question.

**Comment by SimonSapin** [comment]
Oh, I just realized that WeasyPrint has `bookmark-label` based on an earlier version of the spec, with keywords like `content-element`, whereas the spec now has a function like `content(text)`.

Also, what I suggest above doesn’t work at all! Page-margin boxes are generated for every page only after the content of the entire document has been laid out. (This is to support `counter(pages)`, inserting the total number of pages in the document.) So every page would get the value of named strings as they are at the end of the document. So… yeah, that doesn’t work. I also don’t know how to deal with the second argument of the `string()` function. I suppose every box could have its own copy of the mapping of all named strings in the document… I’m not sure exactly how that would work, but this is where you can get creative! So much duplication also sounds quite wasteful, but WeasyPrint already tends to be pretty wasteful with both CPU time and memory usage anyway.

By the way, this part of the original message made me chuckle:
> It's quite simple to add.

**Comment by E-M-P-I-R-E** [comment]
The idea of me contributing to CSS specification makes me chuckle. I mean, that would be awesome but what’s next... Kozea’s ambassador to the United Nations??

Jokes aside... I’m not sure what you mean by this:

>Also, what I suggest above doesn’t work at all! 

From your previous comment:

>it’s not clear to me whether the `string()` function of the `content` property is valid in contexts other than on page-margin boxes (e.g. `::before` or `::after` pseudo-elements) and what the second argument means in this case.

Page-margin boxes, and the first [required] argument to `string-set` are my immediate goal.

I guess, the problem you were getting at is that the page-margin boxes’ content cannot be changed at the time that bookmark-names [where the values _could_ come from] are determined, because the layout is complete?

> box generation is not aware of pages

`page` could be added as an optional argument along with `context` as you suggested [`make_margin_boxes()` -> `content_to_boxes()`]. Side-note: `string()` can only be affected by elements on the same or preceding pages right?

> So every page would get the value of named strings as they are at the end of the document.

I don’t see why this is necessarily the case - say, all the required information was available?


Do you foresee major changes to the specification or the areas in WeasyPrint that will be affected by adding this feature? These comments might be relevant:

>`# If this is changed, maybe update weasy.layout.pages.make_margin_boxes()`

and in that function

>`# TODO: get actual counter values at the time of the last page break`

I’d love to contribute and continue to learn from this project. I’ve definitely made some progress, at least insofar as my own understanding.

The question still remains - at what point should the data for named-strings be populated and where should it be stored.

Thanks!

**Comment by SimonSapin** [comment]
For what it’s worth, you don’t need to represent anyone to provide feedback to specs.

> Jokes aside... I’m not sure what you mean by this:
>> Also, what I suggest above doesn’t work at all!

This refers to the fact that, if you just implement what I described before, you will get incorrect results since the content of page-margin boxes is currently determined after the content of all pages has been laid out, so my suggestion to have `LayoutContext` contain just a map of string name to current value is not good.

> Page-margin boxes, and the first [required] argument to string-set are my immediate goal.

Ok, cool. It’s better to do smaller things at a time and then iterate as needed.

>  Side-note: string() can only be affected by elements on the same or preceding pages right?

No. If you have, say, `h1 { string-set: chapter content(text) } @page { @top-center { content: string(chapter) } }` and a given chapter is many pages long, a given page can have a chapter title that was set many pages before.

>> So every page would get the value of named strings as they are at the end of the document.
> I don’t see why this is necessarily the case

That if you do everything as I suggested in https://github.com/Kozea/WeasyPrint/issues/174#issuecomment-76711593 , because that doesn’t work.

>> `# If this is changed, maybe update weasy.layout.pages.make_margin_boxes()`

I unfortunately don’t remember what this comment refers to :x

>> `# TODO: get actual counter values at the time of the last page break`

This is not directly related to named pages. This is about supporting counters other than `page` or `pages` in page-margin boxes, e.g. this currently doesn’t work: `h1 { counter-increment: chapter } @page { @top-center { content: "Chapter " counter(chapter) "." } }`

> The question still remains - at what point should the data for named-strings be populated and where should it be stored.

Yeah, that’s an important question, and the suggestion I initially made doesn’t work. Unfortunately I can’t spend much time on this… so I look forward to read what you think! :)

**Comment by E-M-P-I-R-E** [comment]
I do have it working for one value-pair in the content-list. I realized that I would need support of the second argument of `string()` so I did get that done. Basically, after a page is generated [in `make_page()`], I traverse the elements and record any new values for the string. `content_to_boxes()` calls my new function `get_string_set_for()` which handles the rules of the second argument and returns the correct value.

I don't know why values other than `text` don't work for the content function. I am calling the same functions in `TEXT_CONTENT_EXTRACTORS` that are used by `resolve_bookmark_labels()`. At one point I replaced `children` with `descendants()` and the functions were seeing value in psuedo-elements (but it was doubled).

I wrote a few tests and one [`test_margin_box_string_set_after()`] is failing for the reason above. Aside from that, I don't think I broke anything [though, there are some tests failing anyways].

https://github.com/E-M-P-I-R-E/WeasyPrint/commit/bc51dec015986565b003a5cb6482f72ba3b1b820

**Comment by E-M-P-I-R-E** [comment]
Okay, I made a pull request. I found a way to support making assignments using the psuedo-elements (`::before`, `::after`) but ended up removing it in my last commit. I think I understand what you were saying here, now.

>This means that string() would not be supported on (pseudo-)elements, but arguably that’s not very useful anyway

If you're setting the value on multiple elements, then those elements would all get the same value for their respective psuedo-element. Plus, the content will be part of the text anyways. More importantly, I was unsure about the change I had to make to the function in `build.TEXT_CONTENT_EXTRACTORS` (although it didn't break any existing tests). You can see what I removed in my final commit (`pages.py` & `build.py`).

I hope it's good for everyone =)!

**Comment by SimonSapin** [comment]
The PR is #246. I’ll comment there.

**Comment by liZe** [comment]
#246 merged.
